### PR TITLE
Buffer flags

### PR DESCRIPTION
**Pull request by juimonen**
Change to use the buffer caps' upper 16 bit to other than solely memory caps. First ones are buffer underrun and overrun flags in bit positions 16 and 17. This way topology changes are minimal, and no changes in driver at all.

### CODE REVIEW

File: src/audio/buffer.c [hunk 1]
```diff
@@ -27,6 +27,7 @@ uintptr_t buffer_uuid_ptr = SOF_UUID(buffer_uuid);
 struct comp_buffer *buffer_alloc(uint32_t size, uint32_t caps, uint32_t align)
 {
 	struct comp_buffer *buffer;
+	uint32_t mem_caps = caps &= ~0xFFFF0000;
```
Comment (xiulipan) [inline src/audio/buffer.c:30]: Why not 
```suggestion
	uint32_t mem_caps = caps &= 0xFFFF;
```

Comment (juimonen) [reply src/audio/buffer.c]: @xiulipan hmm that doesn't work for me.... I guess the hex constants define the number of bits also. So your suggestion will be become 16 bit unsigned, then promoted to 32 bit -> so 16 msb's are still zero.

Comment (juimonen) [reply src/audio/buffer.c]: @xiuli sorry didn't notice that you removed the ~. 

anyway what ever is more clear for people to understand what I'm trying to do here...

File: src/audio/buffer.c [hunk 2]
```diff
@@ -27,6 +27,7 @@ uintptr_t buffer_uuid_ptr = SOF_UUID(buffer_uuid);
 struct comp_buffer *buffer_alloc(uint32_t size, uint32_t caps, uint32_t align)
 {
 	struct comp_buffer *buffer;
+	uint32_t mem_caps = caps & ~0xFFFF0000;
```
Comment (lgirdwood) [inline src/audio/buffer.c:30]: We need a macro for this mask value. It's also worth an inline comment here too.

File: src/include/ipc/topology.h [hunk 3]
```diff
@@ -86,12 +86,12 @@ struct sof_ipc_comp {
 /*
  * overrun will cause ring buffer overwrite, instead of XRUN.
  */
```
Comment (dbaluta) [inline src/include/ipc/topology.h:88]: It is not clear from the commit message why we change the bit positions? What's wrong with the current positions (0 and 1?)

Comment (paulstelian97) [reply src/include/ipc/topology.h:88]: I think the intent is to separate memory specific parameters from those specific to the buffer itself. For example, bits 0-15 could contain memory caps like allocating the buffer from a DMA-able region while 16-31 would contain stuff that matters for topology (overrun/underrun permitted).

Comment (juimonen) [reply src/include/ipc/topology.h:88]: yes, we added "flags" field which was completely unnecessary as we have about 24 bits left in "caps". So the bit position 0 and 1 we're in the other parameter before. If we can use the caps for other parameters (so exactly like the upper 16bit and leave still 8 for possible future mem flags), I don't have to touch topology. Adding the new flags to topology would touch every buffer definition in every topology and I'm paranoid with this kind of changes, as we have screwed up tplg+fw+kernel combinations before....

Comment (juimonen) [reply src/include/ipc/topology.h:88]: So this is a very good example why we need always driver side PR together with fw PR to see the possible topology and other changes before we merge firmware.

Comment (dbaluta) [reply src/include/ipc/topology.h:88]: @paulstelian97 @juimonen thanks for explanation. Can you please add this explanation in the commit message.

Comment (dbaluta) [reply src/include/ipc/topology.h:88]: @juimonen do we also need to bump the tplg abi?

File: src/audio/buffer.c (hunk 1 above)
Comment (xiulipan) [reply src/audio/buffer.c]: @juimonen Try 0xFFFFL to make compiler take this value as 0x0000FFFF

File: src/audio/buffer.c (hunk 2 above)
Comment (xiulipan) [reply src/audio/buffer.c:30]:
```suggestion
	uint32_t mem_caps = caps & 0xFFFFL;
```
This looks more straight forward for only need lower 16bits

Comment (juimonen) [reply src/audio/buffer.c]: added now 0xFFFF in a macro, seems to work.

File: src/include/ipc/topology.h (hunk 3 above)
Comment (juimonen) [reply src/include/ipc/topology.h:88]: @dbaluta hmm I think it should be visible in the commit messages... there's like 5 commits? the PR message might not contain all relevant info. Sorry if I'm wrong here, I'll correct it then... I'm not including my own rants how stupid we sometimes do things in commit messages :)

**Comment by juimonen** [comment]
@xiulipan @lgirdwood @dbaluta tried to address the comments. If something still needed please let me know...

**Comment by dbaluta** [comment]
@juimonen LGTM. Thanks!

**Comment by zrombel** [comment]
Due to recent PR3194 merge, please rebase for TGL Multicore tests to Pass on Internal CI.

**Comment by paulstelian97** [comment]
@juimonen Please send this PR again based on the "main" branch. You can make it out of the same branch on your end.

**Pull request closed without merging**